kernel/src/common/regs/README.md

```diff
@@ -0,0 +1,207 @@
+    cr: ReadWrite<u8, Control>,
+
+    // Status register: read-only
+    s: ReadOnly<u8, Status>
+
+    // Registers can be bytes, halfwords, or words:
+    // Note that the second type parameter can be omitted, meaning that there
+    // are no bitfields defined for these registers.
+    byte0: ReadWrite<u8>,
+    byte1: ReadWrite<u8>,
+    short: ReadWrite<u16>,
+    word: ReadWrite<u32>
+
+    // Etc.
+}
+```
+
+## Defining bitfields
+
+Bitfields are defined through the `bitfields!` macro:
+
+```rust
+bitfields! [
+    // First parameter is the register width for the bitfields. Can be u8, u16,
+    // or u32.
+    u8,
+
+    // Each subsequent parameter is a register abbreviation, its descriptive
+    // name, and its associated bitfields.
+    // The descriptive name defines this 'group' of bitfields. Only registers
+    // defined as ReadWrite<_, Control> can use these bitfields.
+    CR Control [
+        // Bitfields are defined as:
+        // name (Mask(mask), shift) [ /* optional named values */ ]
+
+        // This is a two-bit field which includes bits 4 and 5
+        RANGE (Mask(0b11), 4) [
+            // Each of these defines a name for a value that the bitfield can be
+            // written with or matched against. Note that this set is not exclusive--
+            // the field can still be written with arbitrary constants.
+            VeryHigh = 0,
+            High = 1,
+            Low = 2
+        ],
+
+        // A common case is single-bit bitfields, which usually just mean
+        // 'enable' or 'disable' something. This syntax is a shorthand, defining
+        // EN as a field over bit 3, and INT as a field over bit 2.
+        EN 3 [],
+        INT 2 []
+    ],
+
+    // Without the explanatory comments like above, bitfield definition is quite compact:
+    // Status register
+    S Status [
+        TXCOMPLETE 0 [],
+        TXINTERRUPT 1 [],
+        RXCOMPLETE 2 [],
+        RXINTERRUPT 3 [],
+        MODE (Mask(0b11), 4) [
+            FullDuplex = 0,
+            HalfDuplex = 1,
+            Loopback = 2,
+            Disabled = 3
+        ],
+        ERRORCOUNT (Mask(0b11), 6) []
+    ]
+]
+```
+
+
+## Example: Using registers and bitfields
+
+Assuming we have defined a `Registers` struct and the corresponding bitfields as
+in the previous two sections. We also have an immutable reference to the
+`Registers` struct, named `regs`.
+
+```rust
+// -----------------------------------------------------------------------------
+// RAW ACCESS
+// -----------------------------------------------------------------------------
+
+// Get or set the raw value of the register directly. Nothing fancy:
+regs.cr.set(regs.cr.get() + 1);
+
+
+// -----------------------------------------------------------------------------
+// READ
+// -----------------------------------------------------------------------------
+
+// `range` will contain the unshifted value of the RANGE field, e.g. 0, 1, 2, or 3.
```

Comment:
"unshifted" is confusing here. Does that mean it is unshifted from its spot in the register, or that it is no longer shifted to its actual spot in the register?

Reply:
"Unshifted" here means unshifted from its spot in the register. If I just remove the word 'unshifted', is the meaning clear?